source/MRMesh/MRSharpenMarchingCubesMesh.cpp

```diff
@@ -0,0 +1,132 @@
+#include "MRSharpenMarchingCubesMesh.h"
+#include "MRMesh.h"
+#include "MRRingIterator.h"
+#include "MRBitSetParallelFor.h"
+#include "MRBestFit.h"
+#include "MRGeodesicPath.h"
+
+namespace MR
+{
+
+void sharpenMarchingCubesMesh( const Mesh & ref, Mesh & vox, Vector<VoxelId, FaceId> & face2voxel,
+    const SharpenMarchingCubesMeshSettings & settings )
+{
+    Vector<Vector3f, VertId> normals( vox.topology.vertSize() );
+    // find normals and correct points
+    tbb::parallel_for( tbb::blocked_range<VertId>( 0_v, normals.endId() ), [&] ( const tbb::blocked_range<VertId>& range )
+    {
+        for ( VertId v = range.begin(); v < range.end(); ++v )
+        {
+            if ( !vox.topology.hasVert( v ) )
+                continue;
+            const auto proj = findProjection( vox.points[v], ref );
+
+            Vector3f n = ( vox.points[v] - proj.proj.point ).normalized();
+            Vector3f np = ref.pseudonormal( proj.mtp );
```

Comment:
I've tried `ref.normal(ref.topology.left(proj.mtp.e))` and sometimes it was better